auth_mellon_util.c

```diff
@@ -1835,7 +2147,127 @@ char *am_get_assertion_consumer_service_by_binding(LassoProvider *provider, cons
 
+/* Join array of strings with separator
+ *
+ * Parameters:
+ *  pool      memory allocation pool
+ *  strings   NULL terminated array of NULL terminated strings
+ *  separator string to insert between elements
+ *
+ * Example:
+ *  char *strings[] = {"one", "two", NULL};
+ *  am_str_join(pool, strings, ", ")
+ *
+ *  would return "one, two"
+ */
+
+char *am_str_join(apr_pool_t *pool, const char **strings, const char *separator)
+{
+    char *str = NULL;
+    char *dst = NULL;
+    const char **p;
+    apr_size_t i;
+    apr_size_t len;
+    apr_size_t n_strings = 0;
+    apr_size_t total_string_len = 0;
+    apr_size_t separator_len = strlen(separator);
+
+    /*
+     * Iterate over string array, compute total number of characters
+     * and a count of how many strings
+     */
+    for (p = strings; *p; p++) {
+        total_string_len += strlen(*p);
+        n_strings++;
+    }
+
+    /* Add in the length of all the separators */
+    if (n_strings > 0) {
+        total_string_len += separator_len * (n_strings - 1);
+    }
+
+    str = apr_palloc(pool, total_string_len + 1);
+    for (i = 0, dst = str; i < n_strings; i++) {
+        /* If this is not the first string insert a separator */
+        if (i > 0) {
+            memcpy(dst, separator, separator_len);
+            dst += separator_len;
+        }
+        /* Copy the string */
+        len = strlen(strings[i]);
+        memcpy(dst, strings[i], len);
+        dst += len;
+
+    }
+    *dst++ = 0;                 /* NULL terminate */
+
+    return str;
+}
+
 #ifdef HAVE_ECP
+
+/* String representation of ECPServiceOptions bitmask
+ *
+ * ECPServiceOptions is a bitmask of flags. Return a comma separated string
+ * of all the flags. If any bit in the bitmask is unaccounted for an
+ * extra string will be appended of the form "(unknown bits = x)".
+ *
+ * Parameters:
+ *  pool    memory allocation pool
+ *  options bitmask of PAOS options
+ */
+char *am_ecp_service_options_str(apr_pool_t *pool, ECPServiceOptions options)
+{
+    int i = 0;
+    const char *names[10]; /* Warning, increase this if flags are added,
+                              must be n flags + 1 for NULL terminator */
+
+    if (options & ECP_SERVICE_OPTION_CHANNEL_BINDING) {
+        names[i++] = "channel-binding";
+        options &= ~ECP_SERVICE_OPTION_CHANNEL_BINDING;
+    }
+
+    if (options & ECP_SERVICE_OPTION_HOLDER_OF_KEY) {
+        names[i++] = "holder-of-key";
+        options &= ~ECP_SERVICE_OPTION_HOLDER_OF_KEY;
+    }
+
+    if (options & ECP_SERVICE_OPTION_WANT_AUTHN_SIGNED) {
+        names[i++] = "want-authn-signed";
+        options &= ~ECP_SERVICE_OPTION_WANT_AUTHN_SIGNED;
+    }
+
+    if (options & ECP_SERVICE_OPTION_DELEGATION) {
+        names[i++] = "delegation";
+        options &= ~ECP_SERVICE_OPTION_DELEGATION;
+    }
+
+    if (options) {
+        names[i++] = apr_psprintf(pool, "(unknown bits = %#x)", options);
+    }
+
+    names[i] = NULL;
+    return am_str_join(pool, names, ", ");
```

Comment:
After realizing that there is a [`apr_array_pstrcat`](https://apr.apache.org/docs/apr/2.0/group__apr__tables.html#ga6232817e4a32ee5619c99fe55e145c7a)-function in APR, I think this function could be rewritten using that function. I think that would be better than rolling our own version of that function. It would mean that the elements would only be separated by `','`, but I don't think that is a problem.